independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java

```diff
@@ -945,6 +945,10 @@ public PARAM extractParameterInfo(ClassInfo currentClassInfo, ClassInfo actualEn
                 builder.setOptional(true);
             } else if (convertible) {
                 throw new RuntimeException("Invalid parameter type '" + pt + "' used on method " + errorLocation);
+            } else {
+                elementType = toClassName(pt.arguments().get(0), currentClassInfo, actualEndpointInfo, index);
```

Comment:
The next line is fine, but here you're saying that for a parameter of type `Foo<X>` the `elementType` will be `X` and I'm not sure that's true at all, no?

Reply:
Well, that's the whole point of the PR. It guarantees that we have both types at runtime so the `MessageBodyReader` can know this information (the fact that the element type was not being set properly was the cause of the original issue)

Reply:
My point is that IMO the element type is not `X` but `Foo<X>`, or `Foo` in the erased case. I don't see how `X` can be considered an element type since it's not unwrapped.

Reply:
We already record Foo and the signature itself.
It's just that in this case we were not making use of the signature because X was not being recorded

Reply:
It's also worth noting that this is in line with how we have `List` for example

Reply:
Well, for `List<X>` since it's a container, the element type makes sense to be `X` but this is not a container type. Who even makes use of this element type then? If we pass it to things like message writers then it's wrong: we need to match on `Foo<X>` and not `X`.

Reply:
my brains fails to see this, so if you want to do something different than this PR, I'm fine with that

Reply:
Reading this again, maybe is throwing you off is the name of the variable?
Because all this does down the line, is make RESTEasy Reactive use the signature to determine the proper types.

Reply:
No, what's bothering me is the `pt.arguments().get(0)` (as in `X` if `pt == Foo<X>`) where I think it should either be `pt` (as in `Foo<X>`) or we should set `elementType` to some guard value that indicates this is not actually an element of a collection, if that's what we're using it for.

Reply:
How about now?